### PR TITLE
Migrate knotx dependencies to gradle

### PR DESCRIPTION
**Pull request by Mateusz512**
## Description

Knotx-dependencies is now `mvn` based

## Motivation and Context

We want to include knotx-dependencies in composite gradle build

### Maven support
Checked with a simple Maven configuration:
```
<project>
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.mycompany.app</groupId>
  <artifactId>my-app</artifactId>
  <version>1</version>
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>io.knotx</groupId>
        <artifactId>knotx-dependencies</artifactId>
        <version>2.1.1-SNAPSHOT</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
  <dependencies>
    <dependency>
      <groupId>io.vertx</groupId>
      <artifactId>vertx-core</artifactId>
    </dependency>
  </dependencies>
</project>
```
Outputs:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.982 s
[INFO] Finished at: 2020-02-04T09:18:51+01:00
[INFO] ------------------------------------------------------------------------
```
### Release validation
Screenshot 2020-02-04 at 15 48 28

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,23 +4,10 @@
...
-        <groupId>io.knotx</groupId>
-        <artifactId>knotx-dependencies</artifactId>
-        <version>${knotx.version}</version>
-        <type>pom</type>
-        <scope>import</scope>
-      </dependency>
-    </dependencies>
-  </dependencyManagement>
+```kotlin
+implementation(platform("io.knotx:knotx-dependencies:${project.version}"))
```
Comment (tomaszmichalak) [inline README.md:10]: Do we still support Maven?

File: README.md [hunk 2]
```diff
@@ -4,23 +4,10 @@
...
-      <dependency>
-        <groupId>io.knotx</groupId>
-        <artifactId>knotx-dependencies</artifactId>
-        <version>${knotx.version}</version>
-        <type>pom</type>
-        <scope>import</scope>
-      </dependency>
-    </dependencies>
-  </dependencyManagement>
+```kotlin
```
Comment (tomaszmichalak) [inline README.md:9]: Please add entries in CHANGELOG.md.

File: README.md [hunk 3]
```diff
@@ -4,23 +4,10 @@
 
 ## Use in your project
```
Comment (tomaszmichalak) [inline README.md:5]: Can you please add some description about:
- what is the purpose of this module (see [BOM](https://en.wikipedia.org/wiki/Bill_of_materials), it defines external dependencies versions)
- how to use it in implemented modules (Gradle / Maven)

File: azure-pipelines.yml [hunk 4]
```diff
@@ -76,4 +77,4 @@ steps:
         $(workspaceDir)/$REPO/gradlew -p $(workspaceDir)/$REPO clean test jacocoTestReport
         bash <(curl -s https://codecov.io/bash) -s $(workspaceDir)/$REPO -t $(CODECOV_TOKEN)
       fi
-    displayName: "Run Codecov inspection"
\ No newline at end of file
+    displayName: "Run Codecov inspection"
```
Comment (tomaszmichalak) [inline azure-pipelines.yml]: Not sure, but I think we can remove codecov from this repository.

File: build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,156 @@
+/*
```
Comment (tomaszmichalak) [inline build.gradle.kts:1]: Please remove comments.

File: build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+ * This file was generated by the Gradle 'init' task.
+ */
+
+plugins {
+  `java-platform`
+  `maven-publish`
+}
+
+defaultTasks("publishKnotxPublicationToMavenLocal")
```
Comment (tomaszmichalak) [inline build.gradle.kts]: Is it required?

File: build.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+tasks {
+  getByName("build").dependsOn("publishKnotxPublicationToMavenLocal")
```
Comment (tomaszmichalak) [inline build.gradle.kts]: Is it required?

File: build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+repositories {
+  jcenter()
+  mavenLocal()
```
Comment (tomaszmichalak) [inline build.gradle.kts]: Please remove a local maven repository.

File: build.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+  maven(url = "https://oss.sonatype.org/content/repositories/snapshots")
```
Comment (tomaszmichalak) [inline build.gradle.kts]: Please remove a snapshots repository.

File: build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+  maven(url = "http://repo.maven.apache.org/maven2")
+}
+
+group = "io.knotx"
+version = "2.1.1-SNAPSHOT"
+description = "Knot.x Dependencies - Knot.x components and required dependencies versions"
```
Comment (tomaszmichalak) [inline build.gradle.kts]:
```suggestion
description = "Knot.x Dependencies - BOM containing external Knot.x dependencies versions."
```

File: build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+javaPlatform {
+  allowDependencies()
+}
+
+buildscript {
+  extra.apply {
+    set("vertx.stack.version", "3.8.4")
```
Comment (tomaszmichalak) [inline build.gradle.kts]: Do we need to set it here?

File: build.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    api("org.mockito:mockito-core:${ext("mockito2.version")}")
+    api("org.mockito:mockito-junit-jupiter:${ext("mockito2.version")}")
+    api("com.github.tomakehurst:wiremock:${ext("wiremock.version")}")
+    api("org.junit.platform:junit-platform-surefire-provider:${ext("junit5-surefire.version")}")
+    api("me.alexpanov:free-port-finder:${ext("free-port-finder.version")}")
+    api("org.assertj:assertj-core:${ext("assertj.version")}")
+  }
+  api(platform("io.vertx:vertx-dependencies:${ext("vertx.stack.version")}"))
+  api(platform("org.junit:junit-bom:${ext("junit5.version")}"))
+  api("com.github.stefanbirkner:system-rules:${ext("system-rules.version")}") {
```
Comment (tomaszmichalak) [inline build.gradle.kts]: Can you please check if it is still used?

File: build.gradle.kts [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+  }
+}
+
+publishing {
+  publications {
+    create<MavenPublication>("Knotx") {
+
+      pom {
+        name.set(project.description)
+        description.set("Knot.x stack dependencies POM")
```
Comment (tomaszmichalak) [inline build.gradle.kts]:
```suggestion
        description.set("BOM containing external Knot.x dependencies versions.")
```

File: build.gradle.kts [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+          developerConnection.set("scm:git:ssh://github.com:Knotx/knotx-dependencies.git")
+          url.set("http://github.com/Knotx/knotx-dependencies/tree/master")
+          tag.set("HEAD")
+        }
+
+        withXml {
+          asNode().apply {
+            appendNode("distributionManagement")
+                .apply {
+                  appendNode("snapshotRepository")
```
Comment (tomaszmichalak) [inline build.gradle.kts]: I think a snapshot repository can be removed.

File: build.gradle.kts [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+                      .apply {
+                        appendNode("id", "ossrh")
+                        appendNode("url", "https://oss.sonatype.org/content/repositories/snapshots")
+                      }
+                  appendNode("repository")
+                      .apply {
+                        appendNode("id", "ossrh")
+                        appendNode("url", "https://oss.sonatype.org/content/repositories/snapshots")
```
Comment (tomaszmichalak) [inline build.gradle.kts]:
```suggestion
                        appendNode("url", "https://oss.sonatype.org/service/local/staging/deploy/maven2/")
```

File: build.gradle.kts [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+                      }
+                }
+
+            appendNode("repositories")
```
Comment (tomaszmichalak) [inline build.gradle.kts]: I think it can be skipped.

File: settings.gradle.kts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+/*
```
Comment (tomaszmichalak) [inline settings.gradle.kts:1]: Please remove comments.

File: build.gradle.kts (hunk 5 above)
Comment (tomaszmichalak) [inline build.gradle.kts:1]: Please add RAT plugin to check if all files contain license headers.

**Review by tomaszmichalak — CHANGES_REQUESTED** [review]

File: build.gradle.kts [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+
+repositories {
+  jcenter()
+  mavenLocal()
+  maven(url = "https://oss.sonatype.org/content/repositories/snapshots")
+  maven(url = "http://repo.maven.apache.org/maven2")
+}
+
+group = "io.knotx"
+version = "2.1.1-SNAPSHOT"
```
Comment (malaskowski) [inline build.gradle.kts]: Can we move `version` to `gradle.properties` as we do usually in other projects?

File: build.gradle.kts [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+  api(platform("io.vertx:vertx-dependencies:${ext("vertx.stack.version")}"))
+  api(platform("org.junit:junit-bom:${ext("junit5.version")}"))
+  api("com.github.stefanbirkner:system-rules:${ext("system-rules.version")}") {
+    exclude(group = "junit", module = "junit-dep")
+  }
+}
+
+publishing {
+  publications {
+    create<MavenPublication>("Knotx") {
```
Comment (malaskowski) [inline build.gradle.kts]: Maybe we could use `io.knotx.maven-publish` plugin here?

File: README.md (hunk 1 above)
Comment (Mateusz512) [reply README.md:10]: I don't think we should, there will be two files to maintain.
What are your thoughts on this @tomaszmichalak ?

File: build.gradle.kts (hunk 6 above)
Comment (Mateusz512) [reply build.gradle.kts]: I've changed it to `publishToMavenLocal` so it can be executed with simple `./gradlew`

File: build.gradle.kts (hunk 11 above)
Comment (Mateusz512) [reply build.gradle.kts]: moved to `gradle.properties`

**Review by tomaszmichalak — APPROVED** [review]

**Pull request merged**